### ISSUE TITLE
Trigger violation on mutable values and function calls for non-dataclass attribute defaults

### ISSUE THREAD

**Issue by adampauls** [astral-sh/ruff#4053]
This is a feature request. I see in [this PR](https://github.com/charliermarsh/ruff/pull/3877) that there is some recent work in flagging mutable and function call defaults for dataclass attributes. Is there a plan to enable this check for non-dataclass attributes  as well? Right now, no Python tooling I know of catches this bug:

```python
import dataclasses

class NonDataClass:
    my_list: list[int] = dataclasses.field(default_factory=list)  # oops! type checks but is nonsense

    def add_item(self, item: int) -> None:
        self.my_list.append(item)

instance = NonDataClass()
instance.add_item(1)
assert instance.my_list == [1]
```

The problem is that `dataclass.field` [lies](https://github.com/microsoft/pyright/issues/4980) about its type. 

It would be great if we could check for mutable and function call defaults in class attributes because they are just problematic for non-dataclasses as they are for dataclasses. The only difference from the existing implementation would need to be that `dataclasses.field` should only be exempted when used inside a dataclass. 

I can look into this if the work is not already planned since I think it's a simple change from what's already been done. 

Thanks for the great tool!

**Comment by charliermarsh** [comment]
Interesting! I initially assumed that the suggestion here was to flag usages of `dataclasses.field` within classes that lack the `@dataclass` decorator, but I think you're suggesting something a bit broader: disallowing mutable defaults and function calls within non-`@dataclass`-decorated classes, which would by definition include usages of `dataclasses.field`. Is that right?

**Comment by hmc-cs-mdrissi** [comment]
I'll note some libraries do re-use/support dataclasses field for there own decorator/class. field just serves as a convenient way to hold some annotations and there are number of dataclass like libraries. I do use internal library that does,

```python
from dataclasses import field

@configurable
class Foo:
  x = field(default_factory=list)
```

where default_factory list can be safe to use. It's own separate rule to disallow field in non dataclass sounds fine though as if you aren't using alternative dataclass like libraries it makes sense.

**Comment by adampauls** [comment]
If it's easy, it would be great to allow a configurable list of decorators that allow data class.field, or maybe even any dataclass_transform should be exempted?

**Comment by adampauls** [comment]
> Is that right?

Yes, exactly.

**Comment by adampauls** [comment]
Though I would settle for the more targeted warning on just dataclass.field if that's preferable!

**Comment by charliermarsh** [comment]
Closing in favor of #4390.
